pkg/daemon/server/service/pipeline_metrics_query.go

```diff
@@ -225,7 +225,15 @@ func (ps *pipelineMetadataQuery) GetVertexMetrics(ctx context.Context, req *daem
 		}
 	}
 
-	resp.VertexMetrics = metricsArr
+	var metricsArrFilter []*daemon.VertexMetrics
+
+	// in case any vertex queried for metrics is not running, this check prevents daemon crashes
+	for _, v := range metricsArr {
+		if v != nil {
+			metricsArrFilter = append(metricsArrFilter, v)
```

Comment:
This is like we do following in line 172, and use `append` to add items.

```
var metricsArr []*daemon.VertexMetrics
```

With this approach, if for some reason, one of the reduce pods fails to get data, then the returned list will miss and skip it, thus it makes the items displayed on the UI incorrect. Can we still return something special in this case, and the UI will determine what to display based on the response, so that it does not lose the position on the UI? For example, set `Pendings`  to -1, and empty `ProcessingRates`.

Reply:
so like I mentioned above, when 1 reduce pod misses the data, the graph shows processing rate for remaining pods and it becomes un-clickable

Reply:
The concern here is not about the UI, but the API itself, the API is not returning the correct data.

Let's assume there are 4 partitions (0, 1, 2, 3) for a reduce vertex (thus there are 4 pods), if the pod 2 has an issue for whatever reason, and no metrics is retrieved. The API expects to see the returned metrics list contains items ordered by the index, however, with this approach, only the data of 0, 1, 3 returned, and the client has no idea which one is missing. On the UI, we have some magic not to display any information at all if we detect the data is incomplete, but it would be misinterpreted by other clients (even though we don't have other clients today).

Reply:
i modified the api to return empty processing rates for partitions that are not available. This way the api has data of all partitions for the clients to act upon.